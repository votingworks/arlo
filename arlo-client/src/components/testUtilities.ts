import { createLocation, createMemoryHistory } from 'history'
import { match as routerMatch } from 'react-router-dom'

type MatchParameter<Params> = { [K in keyof Params]?: string }

const generateUrl = <Params extends MatchParameter<Params>>(
  path: string,
  params: Params
): string => {
  let tempPath = path

  for (const param in params) {
    /* istanbul ignore else */
    if (params.hasOwnProperty(param)) {
      const value = params[param]
      tempPath = tempPath.replace(`:${param}`, value as NonNullable<
        typeof value
      >)
    }
  }

  return tempPath
}

export const routerTestProps = <Params extends MatchParameter<Params> = {}>(
  path: string,
  params: Params
) => {
  const match: routerMatch<Params> = {
    isExact: false,
    path,
    url: generateUrl(path, params),
    params,
  }
  const history = createMemoryHistory()
  const location = createLocation(match.url)

  return { history, location, match }
}

/** Credit to https://stackoverflow.com/a/56452779 for solution to mocking React Router props */

export const asyncForEach = async <T>(
  array: T[],
  callback: (value: T, index: number, array: T[]) => Promise<void>
) => {
  for (let index = 0; index < array.length; index++) {
    await callback(array[index], index, array)
  }
}

export default {
  asyncForEach,
  routerTestProps,
}
