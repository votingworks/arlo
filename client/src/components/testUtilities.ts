import React from 'react'
import { RenderResult, act, render } from '@testing-library/react'
import { createLocation, createMemoryHistory } from 'history'
import { match as routerMatch } from 'react-router-dom'

type MatchParameter<Params> = { [K in keyof Params]?: string }

const generateUrl = <Params extends MatchParameter<Params>>(
  path: string,
  params: Params
): string => {
  let tempPath = path

  for (const param in params) {
    /* istanbul ignore else */
    if (Object.prototype.hasOwnProperty.call(params, param)) {
      const value = params[param]
      tempPath = tempPath.replace(`:${param}`, value as NonNullable<
        typeof value
      >)
    }
  }

  return tempPath
}

export const routerTestProps = <Params extends MatchParameter<Params> = {}>(
  path: string,
  params: Params
) => {
  const match: routerMatch<Params> = {
    isExact: false,
    path,
    url: generateUrl(path, params),
    params,
  }
  const history = createMemoryHistory()
  const location = createLocation(match.url)

  return { history, location, match }
}

/** Credit to https://stackoverflow.com/a/56452779 for solution to mocking React Router props */

export const regexpEscape = (s: string) => {
  /* eslint-disable no-useless-escape */
  return s.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&')
}

export const asyncActRender = async (
  component: React.ReactElement
): Promise<RenderResult> => {
  let result: RenderResult
  await act(async () => {
    result = render(component)
  })
  return result!
}

export default {
  routerTestProps,
}
