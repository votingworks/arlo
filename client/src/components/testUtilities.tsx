import React from 'react'
import { createLocation, createMemoryHistory, MemoryHistory } from 'history'
import { match as routerMatch, Router } from 'react-router-dom'
import equal from 'fast-deep-equal'
import { render } from '@testing-library/react'

type MatchParameter<Params> = { [K in keyof Params]?: string }

const generateUrl = <Params extends MatchParameter<Params>>(
  path: string,
  params: Params
): string => {
  let tempPath = path

  for (const param in params) {
    /* istanbul ignore else */
    if (Object.prototype.hasOwnProperty.call(params, param)) {
      const value = params[param]
      tempPath = tempPath.replace(`:${param}`, value as NonNullable<
        typeof value
      >)
    }
  }

  return tempPath
}

/** Credit to https://stackoverflow.com/a/56452779 for solution to mocking React Router props */

export const routerTestProps = <Params extends MatchParameter<Params> = {}>(
  path: string,
  params: Params
) => {
  const match: routerMatch<Params> = {
    isExact: false,
    path,
    url: generateUrl(path, params),
    params,
  }
  const history = createMemoryHistory()
  const location = createLocation(match.url)

  return { history, location, match }
}

// Copied from https://testing-library.com/docs/example-react-router
export const renderWithRouter = (
  ui: React.ReactElement,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  }: { route?: string; history?: MemoryHistory } = {}
) => {
  const Wrapper: React.FC = ({ children }: { children?: React.ReactNode }) => (
    <Router history={history}>{children}</Router>
  )

  return {
    ...render(ui, { wrapper: Wrapper }),
    // Adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  }
}

// withMockFetch is a helper to mock calls to external APIs (e.g. the Arlo backend).
// - It takes an array of expected FetchRequests and a test runner function
// - It mocks window.fetch to return the given response for each request
// - After running the test function, it checks that the actual received
//   requests exactly match the expected requests.

interface FetchRequest {
  url: string
  options?: RequestInit
  response: object | null
  skipBody?: boolean
  error?: {
    status: number
    statusText: string
  }
}

export const withMockFetch = async (
  requests: FetchRequest[],
  testFn: () => Promise<void>
) => {
  const requestsLeft = [...requests]
  const mockFetch = jest.fn(async (url: string, options?: RequestInit) => {
    const [expectedRequest] = requestsLeft.splice(0, 1)
    if (
      expectedRequest &&
      expectedRequest.url === url
      // equal(expectedRequest.options, options)
      // (expectedRequest.skipBody || equal(expectedRequest.options, options))
    ) {
      if (
        expectedRequest.options &&
        expectedRequest.options.body instanceof FormData &&
        options &&
        options.body &&
        options.body instanceof FormData
      ) {
        const expectedData = expectedRequest.options.body
        const receivedData = options.body
        // defined expectedData and receivedData here because TS won't recognize the typechecks above
        // when expectedRequest.options.body and options.body are used in the loop
        const fileNamesMatch: boolean = Array.from(expectedData.keys())
          .map(key => {
            const expectedFile = expectedData.get(key) as File // the only time we ever send a FormData is when submitting a File
            const receivedFile = receivedData.get(key) as File
            if (expectedFile.name === receivedFile.name) {
              return true
            }
            return false
          })
          .every(v => v) // treating it as a list even though we only send one at a time because we don't know the key that will be used
        if (fileNamesMatch)
          return new Response(JSON.stringify(expectedRequest.response))
      } else if (equal(expectedRequest.options, options)) {
        return expectedRequest.error
          ? new Response(
              JSON.stringify(expectedRequest.response),
              expectedRequest.error
            )
          : new Response(JSON.stringify(expectedRequest.response))
      }
    }

    if (expectedRequest) {
      const requestIndex = requests.length - requestsLeft.length - 1
      // eslint-disable-next-line no-console
      console.error(
        `Expected fetch request (${requestIndex}):\n`,
        expectedRequest,
        '\nActual fetch request:\n',
        { url, options }
      )
    } else {
      // eslint-disable-next-line no-console
      console.error('Unexpected extra fetch request:\n', { url, options })
    }
    return new Response(JSON.stringify({}))
  })
  window.fetch = mockFetch as typeof window.fetch

  await testFn()

  const actualRequests = mockFetch.mock.calls.map(([url, options]) => ({
    url,
    options,
  }))
  const expectedRequests = requests.map(({ url, options }) => ({
    url,
    options,
  }))
  expect(actualRequests).toEqual(expectedRequests)
}

export const regexpEscape = (s: string) => {
  /* eslint-disable no-useless-escape */
  return s.replace(/[-\/\\^$*+?.()|[\]{}]/g, '\\$&')
}

export default {
  routerTestProps,
}
