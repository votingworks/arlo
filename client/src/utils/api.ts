import { AxiosRequestConfig } from 'axios'

const parseCookies = () =>
  Object.fromEntries(
    document.cookie.split(';').map(pair => pair.trim().split('='))
  )

export const addCSRFToken = (options?: RequestInit | AxiosRequestConfig) => {
  const token = parseCookies()._csrf_token
  if (
    token &&
    options &&
    ['POST', 'PUT', 'PATCH', 'DELETE'].includes(options.method!)
  )
    return {
      ...options,
      headers: { ...options.headers, 'X-CSRFToken': token },
    }
  return options
}

export const tryJson = (responseText: string) => {
  try {
    return JSON.parse(responseText)
  } catch (err) {
    return {}
  }
}
export class ApiError extends Error {
  public statusCode: number

  public constructor(message: string, statusCode: number) {
    super(message)
    this.statusCode = statusCode
  }
}

export const fetchApi = async (url: string, options?: RequestInit) => {
  const response = await fetch(url, addCSRFToken(options))
  if (response.ok) return response.json()
  const text = await response.text()
  const { errors } = tryJson(text)
  const error = errors && errors.length && errors[0].message
  throw new ApiError(error || text, response.status)
}
