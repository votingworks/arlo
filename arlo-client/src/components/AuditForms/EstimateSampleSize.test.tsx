import React from 'react'
import { render, fireEvent, wait } from '@testing-library/react'
import { toast } from 'react-toastify'
import EstimateSampleSize, {
  TwoColumnSection,
  InputLabelRow,
  InputFieldRow,
  //FieldLeft, TODO: need to mock Formik Field to test these like this
  //FieldRight,
  InputLabel,
  InputLabelRight,
  Action,
} from './EstimateSampleSize'
import { asyncForEach } from '../testUtilities'
import statusStates from './_mocks'
import api from '../utilities'

const apiMock = api as jest.Mock<ReturnType<typeof api>, Parameters<typeof api>>

jest.mock('../utilities')

const estimateSampleSizeMocks = {
  inputs: [
    { key: 'audit-name', value: 'Election Name' },
    { key: 'contest-1-name', value: 'Contest Name' },
    { key: 'contest-1-choice-1-name', value: 'Choice One' },
    { key: 'contest-1-choice-2-name', value: 'Choice Two' },
    { key: 'contest-1-choice-1-votes', value: '10' },
    { key: 'contest-1-choice-2-votes', value: '20' },
    { key: 'contest-1-total-ballots', value: '30' },
    { key: 'risk-limit', value: '2' },
    { key: 'random-seed', value: '12345678901234512345' },
  ],
  errorInputs: [
    { key: 'audit-name', value: '', error: 'Required' },
    { key: 'contest-1-name', value: '', error: 'Required' },
    { key: 'contest-1-choice-1-name', value: '', error: 'Required' },
    { key: 'contest-1-choice-2-name', value: '', error: 'Required' },
    {
      key: 'contest-1-choice-1-votes',
      value: '',
      error: 'Must be a number',
    },
    {
      key: 'contest-1-choice-1-votes',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'contest-1-choice-1-votes',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'contest-1-choice-1-votes',
      value: '0.5',
      error: 'Must be an integer',
    },
    {
      key: 'contest-1-choice-2-votes',
      value: '',
      error: 'Must be a number',
    },
    {
      key: 'contest-1-choice-2-votes',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'contest-1-choice-2-votes',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'contest-1-choice-2-votes',
      value: '0.5',
      error: 'Must be an integer',
    },
    {
      key: 'contest-1-total-ballots',
      value: '',
      error: 'Must be a number',
    },
    {
      key: 'contest-1-total-ballots',
      value: 'test',
      error: 'Must be a number',
    },
    {
      key: 'contest-1-total-ballots',
      value: '-1',
      error: 'Must be a positive number',
    },
    {
      key: 'contest-1-total-ballots',
      value: '0.5',
      error: 'Must be an integer',
    },
    { key: 'random-seed', value: '', error: 'Required' },
    {
      key: 'random-seed',
      value: 'test',
      error: 'Must be only numbers',
    },
    {
      key: 'random-seed',
      value: '123451234512345123451',
      error: 'Must be 20 digits or less',
    },
  ],
  post: {
    method: 'POST',
    body: {
      name: 'Election Name',
      randomSeed: '12345678901234512345',
      riskLimit: 2,
      contests: [
        {
          id: expect.stringMatching(/^[-0-9a-z]*$/),
          name: 'Contest Name',
          totalBallotsCast: 30,
          choices: [
            {
              id: expect.stringMatching(/^[-0-9a-z]*$/),
              name: 'Choice One',
              numVotes: 10,
            },
            {
              id: expect.stringMatching(/^[-0-9a-z]*$/),
              name: 'Choice Two',
              numVotes: 20,
            },
          ],
        },
      ],
    },
    headers: {
      'Content-Type': 'application/json',
    },
  },
}

function getDisplayName(WrappedComponent: React.ComponentClass) {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component'
}

describe('EstimateSampleSize', () => {
  ;[
    TwoColumnSection,
    InputLabelRow,
    InputFieldRow,
    //FieldLeft, TODO: need to mock Formik Field to test these like this
    //FieldRight,
    InputLabel,
    InputLabelRight,
    Action,
  ].forEach((Component: any) => {
    it(`renders ${getDisplayName(Component)} correctly`, () => {
      const { container } = render(<Component />)
      expect(container).toMatchSnapshot()
    })
  })

  it('renders empty state correctly', () => {
    const { container, rerender } = render(
      <EstimateSampleSize
        audit={statusStates[0]}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        electionId="1"
      />
    )
    expect(container).toMatchSnapshot()

    rerender(
      <EstimateSampleSize
        audit={statusStates[0]}
        isLoading
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        electionId="1"
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('renders after contests creation correctly', () => {
    const { container, rerender } = render(
      <EstimateSampleSize
        audit={statusStates[1]}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        electionId="1"
      />
    )
    expect(container).toMatchSnapshot()

    rerender(
      <EstimateSampleSize
        audit={statusStates[1]}
        isLoading
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        electionId="1"
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('renders after sample size is estimated correctly', () => {
    const { container, rerender } = render(
      <EstimateSampleSize
        audit={statusStates[2]}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        electionId="1"
      />
    )
    expect(container).toMatchSnapshot()

    rerender(
      <EstimateSampleSize
        audit={statusStates[2]}
        isLoading
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        electionId="1"
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('renders after manifest is uploaded correctly', () => {
    const { container, rerender } = render(
      <EstimateSampleSize
        audit={statusStates[3]}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        electionId="1"
      />
    )
    expect(container).toMatchSnapshot()

    rerender(
      <EstimateSampleSize
        audit={statusStates[3]}
        isLoading
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        electionId="1"
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('renders during rounds stage correctly', () => {
    const { container, rerender } = render(
      <EstimateSampleSize
        audit={statusStates[4]}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        electionId="1"
      />
    )
    expect(container).toMatchSnapshot()

    rerender(
      <EstimateSampleSize
        audit={statusStates[4]}
        isLoading
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        electionId="1"
      />
    )
    expect(container).toMatchSnapshot()
  })

  it.skip('adds and removes contests', async () => {
    // skip until feature is complete in backend
    const { getByText, getAllByText, queryByText } = render(
      <EstimateSampleSize
        audit={statusStates[0]}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        electionId="1"
      />
    )

    fireEvent.click(getByText('Add another targeted contest'))

    expect(
      getAllByText('Enter the name of the contest that will drive the audit.')
        .length
    ).toBe(2)
    expect(getByText('Contest 1 Name')).toBeTruthy()
    expect(getByText('Contest 2 Name')).toBeTruthy()

    fireEvent.click(getByText('Remove Contest 2'))

    expect(
      getAllByText('Enter the name of the contest that will drive the audit.')
        .length
    ).toBe(1)
    expect(getByText('Contest Name')).toBeTruthy()
    await wait(() => {
      expect(queryByText('Contest 2')).not.toBeInTheDocument()
      expect(queryByText('Remove Contest 1')).not.toBeInTheDocument()
    })
  })

  it('adds and removes choices', async () => {
    const { getByText, getAllByText, queryAllByText } = render(
      <EstimateSampleSize
        audit={statusStates[0]}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        electionId="1"
      />
    )

    fireEvent.click(getByText('Add a new candidate/choice'), { bubbles: true })

    expect(getAllByText(/Name of Candidate\/Choice \d/i).length).toBe(3)
    expect(getAllByText(/Votes for Candidate\/Choice \d/i).length).toBe(3)
    expect(getAllByText(/Remove choice \d/i).length).toBe(3)

    fireEvent.click(getByText('Remove choice 1'), { bubbles: true })

    await wait(() => {
      expect(queryAllByText(/Remove choice \d/i).length).toBe(0)
      expect(getAllByText(/Name of Candidate\/Choice \d/i).length).toBe(2)
      expect(getAllByText(/Votes for Candidate\/Choice \d/i).length).toBe(2)
    })
  })

  it('is able to submit the form successfully', async () => {
    apiMock.mockImplementation(async () => ({
      message: 'success',
      ok: true,
    }))
    const updateAuditMock = jest.fn()
    const setIsLoadingMock = jest.fn()

    const { getByTestId } = render(
      <EstimateSampleSize
        audit={statusStates[0]}
        isLoading={false}
        setIsLoading={setIsLoadingMock}
        updateAudit={updateAuditMock}
        electionId="1"
      />
    )

    estimateSampleSizeMocks.inputs.forEach(inputData => {
      const input = getByTestId(inputData.key) as HTMLInputElement
      fireEvent.change(input, { target: { value: inputData.value } })
      expect(input.value).toBe(inputData.value)
    })

    fireEvent.click(getByTestId('submit-form-one'), { bubbles: true })
    await wait(() => {
      const { body } = apiMock.mock.calls[0][1] as { body: string }
      expect(setIsLoadingMock).toHaveBeenCalledTimes(2)
      expect(apiMock).toHaveBeenCalledTimes(1)
      expect(apiMock.mock.calls[0][0]).toBe('/audit/basic')
      expect(JSON.parse(body)).toMatchObject(estimateSampleSizeMocks.post.body)
      expect(updateAuditMock).toHaveBeenCalledTimes(1)
    })
  })

  it('displays errors', async () => {
    apiMock.mockReset()
    const { getByTestId } = render(
      <EstimateSampleSize
        audit={statusStates[0]}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={jest.fn()}
        electionId="1"
      />
    )

    await asyncForEach(estimateSampleSizeMocks.errorInputs, async inputData => {
      const { key, value, error } = inputData
      const input = getByTestId(key) as HTMLInputElement
      const errorID = input.name + '-error'
      fireEvent.change(input, { target: { value: value } })
      fireEvent.blur(input)
      await wait(() => {
        expect({
          text: getByTestId(errorID).textContent,
          context: `${key}, ${value}: ${input.value}, ${error}`,
        }).toMatchObject({
          text: error,
          context: `${key}, ${value}: ${input.value}, ${error}`,
        })
      })
    })

    fireEvent.click(getByTestId('submit-form-one'), { bubbles: true })
    await wait(() => {
      expect(apiMock.mock.calls.length).toBe(0) // doesn't post because of errors
    })
  })

  it('handles errors from the form submission', async () => {
    apiMock.mockReset()
    apiMock.mockImplementation(() =>
      Promise.reject({
        message: 'A test error',
        ok: false,
      })
    )
    const toastSpy = jest.spyOn(toast, 'error').mockImplementation()
    const updateAuditMock = jest.fn()
    const { getByTestId } = render(
      <EstimateSampleSize
        audit={statusStates[0]}
        isLoading={false}
        setIsLoading={jest.fn()}
        updateAudit={updateAuditMock}
        electionId="1"
      />
    )

    estimateSampleSizeMocks.inputs.forEach(inputData => {
      const input = getByTestId(inputData.key) as HTMLInputElement
      fireEvent.change(input, { target: { value: inputData.value } })
      expect(input.value).toBe(inputData.value)
    })

    fireEvent.click(getByTestId('submit-form-one'), { bubbles: true })

    await wait(() => {
      expect(apiMock.mock.calls.length).toBe(1)
      expect(toastSpy).toHaveBeenCalledTimes(1)
      expect(toastSpy).toHaveBeenCalledWith('A test error')
      expect(updateAuditMock).toHaveBeenCalledTimes(0)
    })
  })
})
