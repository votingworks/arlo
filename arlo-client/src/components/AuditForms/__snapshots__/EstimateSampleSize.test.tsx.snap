// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EstimateSampleSize renders corretly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form>
        <div
          class="sc-ifAKCX jXnARC"
        >
          <div
            class="sc-bxivhb doxbWk"
          >
            Contest Information
          </div>
          <div
            class="sc-bdVaJa fdbLH"
          >
            <div
              class="sc-htpNat gYCOid"
            >
              Election Name
            </div>
            <div
              class="sc-htoDjs jjofgS"
            >
              <input
                class="sc-dnqmqq dvJRiU"
                disabled=""
                name="name"
                value="Primary 2019"
              />
            </div>
          </div>
          <div
            class="sc-bdVaJa fdbLH"
          >
            <div
              class="sc-htpNat gYCOid"
            >
              Contest  Name
            </div>
            <div
              class="sc-bwzfXH xctbO"
            >
              Enter the name of the contest that will drive the audit.
            </div>
            <div
              class="sc-htoDjs jjofgS"
            >
              <input
                class="sc-dnqmqq dvJRiU"
                disabled=""
                name="contests[0].name"
                value="Contest 1"
              />
            </div>
          </div>
          <div
            class="sc-bdVaJa fdbLH"
          >
            <div
              class="sc-htpNat gYCOid"
            >
              Candidates/Choices & Vote Totals
            </div>
            <div
              class="sc-bwzfXH xctbO"
            >
              Enter the name of each candidate choice that appears on the ballot for this contest.
            </div>
            <div
              class="sc-VigVT gSWWKv"
            >
              <div
                class="sc-jTzLTM kHOUMK"
              >
                <label
                  class="sc-kAzzGY cmOrRM"
                >
                  Name of Candidate/Choice 
                  1
                </label>
                <label
                  class="sc-chPdSV kHjWou"
                >
                  Votes for Candidate/Choice 
                  1
                </label>
              </div>
              <div
                class="sc-fjdhpX gYcVlS"
              >
                <div
                  class="sc-jzJRlG jxSbtE sc-htoDjs jjofgS"
                >
                  <input
                    class="sc-dnqmqq dvJRiU"
                    disabled=""
                    name="contests[0].choices[0].name"
                    value="Candidate 1"
                  />
                </div>
                <div
                  class="sc-jzJRlG sc-cSHVUG cDvPFF sc-htoDjs jjofgS"
                >
                  <input
                    class="sc-dnqmqq dvJRiU"
                    disabled=""
                    name="contests[0].choices[0].numVotes"
                    type="number"
                    value="42"
                  />
                </div>
              </div>
              <div
                class="sc-jTzLTM kHOUMK"
              >
                <label
                  class="sc-kAzzGY cmOrRM"
                >
                  Name of Candidate/Choice 
                  2
                </label>
                <label
                  class="sc-chPdSV kHjWou"
                >
                  Votes for Candidate/Choice 
                  2
                </label>
              </div>
              <div
                class="sc-fjdhpX gYcVlS"
              >
                <div
                  class="sc-jzJRlG jxSbtE sc-htoDjs jjofgS"
                >
                  <input
                    class="sc-dnqmqq dvJRiU"
                    disabled=""
                    name="contests[0].choices[1].name"
                    value="Candidate 2"
                  />
                </div>
                <div
                  class="sc-jzJRlG sc-cSHVUG cDvPFF sc-htoDjs jjofgS"
                >
                  <input
                    class="sc-dnqmqq dvJRiU"
                    disabled=""
                    name="contests[0].choices[1].numVotes"
                    type="number"
                    value="72"
                  />
                </div>
              </div>
            </div>
          </div>
          <div
            class="sc-bdVaJa fdbLH"
          >
            <div
              class="sc-htpNat gYCOid"
            >
              Total Ballots Cast
            </div>
            <div
              class="sc-bwzfXH xctbO"
            >
              Enter the overall number of ballot cards cast in jurisdictions containing this contest.
            </div>
            <div
              class="sc-htoDjs jjofgS"
            >
              <input
                class="sc-dnqmqq dvJRiU"
                disabled=""
                name="contests[0].totalBallotsCast"
                type="number"
                value="4200"
              />
            </div>
          </div>
          <div
            class="sc-bxivhb doxbWk"
          >
            Audit Settings
          </div>
          <div
            class="sc-bdVaJa fdbLH"
          >
            <div
              class="sc-htpNat gYCOid"
            >
              Desired Risk Limit
            </div>
            <div
              class="sc-bwzfXH xctbO"
            >
              Set the risk for the audit as as percentage (e.g. "5" = 5%)
            </div>
            <select
              disabled=""
              name="riskLimit"
            >
              <option>
                1
              </option>
              <option>
                2
              </option>
              <option>
                3
              </option>
              <option>
                4
              </option>
              <option>
                5
              </option>
              <option>
                6
              </option>
              <option>
                7
              </option>
              <option>
                8
              </option>
              <option>
                9
              </option>
              <option>
                10
              </option>
              <option>
                11
              </option>
              <option>
                12
              </option>
              <option>
                13
              </option>
              <option>
                14
              </option>
              <option>
                15
              </option>
              <option>
                16
              </option>
              <option>
                17
              </option>
              <option>
                18
              </option>
              <option>
                19
              </option>
              <option>
                20
              </option>
            </select>
          </div>
          <div
            class="sc-bdVaJa fdbLH"
          >
            <div
              class="sc-htpNat gYCOid"
            >
              Random Seed
            </div>
            <div
              class="sc-bwzfXH xctbO"
            >
              Enter the random number to seed the pseudo-random number generator.
            </div>
            <div
              class="sc-htoDjs jjofgS"
            >
              <input
                class="sc-dnqmqq dvJRiU"
                disabled=""
                name="randomSeed"
                type="number"
                value="123123123"
              />
            </div>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form>
      <div
        class="sc-ifAKCX jXnARC"
      >
        <div
          class="sc-bxivhb doxbWk"
        >
          Contest Information
        </div>
        <div
          class="sc-bdVaJa fdbLH"
        >
          <div
            class="sc-htpNat gYCOid"
          >
            Election Name
          </div>
          <div
            class="sc-htoDjs jjofgS"
          >
            <input
              class="sc-dnqmqq dvJRiU"
              disabled=""
              name="name"
              value="Primary 2019"
            />
          </div>
        </div>
        <div
          class="sc-bdVaJa fdbLH"
        >
          <div
            class="sc-htpNat gYCOid"
          >
            Contest  Name
          </div>
          <div
            class="sc-bwzfXH xctbO"
          >
            Enter the name of the contest that will drive the audit.
          </div>
          <div
            class="sc-htoDjs jjofgS"
          >
            <input
              class="sc-dnqmqq dvJRiU"
              disabled=""
              name="contests[0].name"
              value="Contest 1"
            />
          </div>
        </div>
        <div
          class="sc-bdVaJa fdbLH"
        >
          <div
            class="sc-htpNat gYCOid"
          >
            Candidates/Choices & Vote Totals
          </div>
          <div
            class="sc-bwzfXH xctbO"
          >
            Enter the name of each candidate choice that appears on the ballot for this contest.
          </div>
          <div
            class="sc-VigVT gSWWKv"
          >
            <div
              class="sc-jTzLTM kHOUMK"
            >
              <label
                class="sc-kAzzGY cmOrRM"
              >
                Name of Candidate/Choice 
                1
              </label>
              <label
                class="sc-chPdSV kHjWou"
              >
                Votes for Candidate/Choice 
                1
              </label>
            </div>
            <div
              class="sc-fjdhpX gYcVlS"
            >
              <div
                class="sc-jzJRlG jxSbtE sc-htoDjs jjofgS"
              >
                <input
                  class="sc-dnqmqq dvJRiU"
                  disabled=""
                  name="contests[0].choices[0].name"
                  value="Candidate 1"
                />
              </div>
              <div
                class="sc-jzJRlG sc-cSHVUG cDvPFF sc-htoDjs jjofgS"
              >
                <input
                  class="sc-dnqmqq dvJRiU"
                  disabled=""
                  name="contests[0].choices[0].numVotes"
                  type="number"
                  value="42"
                />
              </div>
            </div>
            <div
              class="sc-jTzLTM kHOUMK"
            >
              <label
                class="sc-kAzzGY cmOrRM"
              >
                Name of Candidate/Choice 
                2
              </label>
              <label
                class="sc-chPdSV kHjWou"
              >
                Votes for Candidate/Choice 
                2
              </label>
            </div>
            <div
              class="sc-fjdhpX gYcVlS"
            >
              <div
                class="sc-jzJRlG jxSbtE sc-htoDjs jjofgS"
              >
                <input
                  class="sc-dnqmqq dvJRiU"
                  disabled=""
                  name="contests[0].choices[1].name"
                  value="Candidate 2"
                />
              </div>
              <div
                class="sc-jzJRlG sc-cSHVUG cDvPFF sc-htoDjs jjofgS"
              >
                <input
                  class="sc-dnqmqq dvJRiU"
                  disabled=""
                  name="contests[0].choices[1].numVotes"
                  type="number"
                  value="72"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="sc-bdVaJa fdbLH"
        >
          <div
            class="sc-htpNat gYCOid"
          >
            Total Ballots Cast
          </div>
          <div
            class="sc-bwzfXH xctbO"
          >
            Enter the overall number of ballot cards cast in jurisdictions containing this contest.
          </div>
          <div
            class="sc-htoDjs jjofgS"
          >
            <input
              class="sc-dnqmqq dvJRiU"
              disabled=""
              name="contests[0].totalBallotsCast"
              type="number"
              value="4200"
            />
          </div>
        </div>
        <div
          class="sc-bxivhb doxbWk"
        >
          Audit Settings
        </div>
        <div
          class="sc-bdVaJa fdbLH"
        >
          <div
            class="sc-htpNat gYCOid"
          >
            Desired Risk Limit
          </div>
          <div
            class="sc-bwzfXH xctbO"
          >
            Set the risk for the audit as as percentage (e.g. "5" = 5%)
          </div>
          <select
            disabled=""
            name="riskLimit"
          >
            <option>
              1
            </option>
            <option>
              2
            </option>
            <option>
              3
            </option>
            <option>
              4
            </option>
            <option>
              5
            </option>
            <option>
              6
            </option>
            <option>
              7
            </option>
            <option>
              8
            </option>
            <option>
              9
            </option>
            <option>
              10
            </option>
            <option>
              11
            </option>
            <option>
              12
            </option>
            <option>
              13
            </option>
            <option>
              14
            </option>
            <option>
              15
            </option>
            <option>
              16
            </option>
            <option>
              17
            </option>
            <option>
              18
            </option>
            <option>
              19
            </option>
            <option>
              20
            </option>
          </select>
        </div>
        <div
          class="sc-bdVaJa fdbLH"
        >
          <div
            class="sc-htpNat gYCOid"
          >
            Random Seed
          </div>
          <div
            class="sc-bwzfXH xctbO"
          >
            Enter the random number to seed the pseudo-random number generator.
          </div>
          <div
            class="sc-htoDjs jjofgS"
          >
            <input
              class="sc-dnqmqq dvJRiU"
              disabled=""
              name="randomSeed"
              type="number"
              value="123123123"
            />
          </div>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
